src/codec.rs

```diff
@@ -1113,71 +1067,110 @@ impl<T: Encode> Encode for [T] {
+	let mut decoded_vec = vec![];
+	let mut num_undecoded_items = len;
+	while num_undecoded_items > 0 {
+		let chunk_len = chunk_len.min(num_undecoded_items);
+		decoded_vec.reserve_exact(chunk_len);
 
-	let input_len = input.remaining_len()?;
+		decode_chunk(&mut decoded_vec, chunk_len)?;
 
-	// If there is input len and it cannot be pre-allocated then return directly.
-	if input_len.map(|l| l < byte_len).unwrap_or(false) {
-		return Err("Not enough data to decode vector".into());
+		num_undecoded_items = num_undecoded_items.saturating_sub(chunk_len);
 	}
 
-	// In both these branches we're going to be creating and resizing a Vec<T>,
-	// but casting it to a &mut [u8] for reading.
-
-	// Note: we checked that if input_len is some then it can preallocated.
-	let r = if input_len.is_some() || byte_len < MAX_PREALLOCATION {
-		// Here we pre-allocate the whole buffer.
-		let mut items: Vec<T> = vec![Default::default(); items_len];
-		let bytes_slice = items.as_mut_byte_slice();
-		input.read(bytes_slice)?;
+	Ok(decoded_vec)
+}
 
-		items
-	} else {
-		// An allowed number of preallocated item.
-		// Note: `MAX_PREALLOCATION` is expected to be more or equal to size of `T`, precondition.
-		let max_preallocated_items = MAX_PREALLOCATION / mem::size_of::<T>();
+/// Create a `Vec<T>` by casting directly from a buffer of read `u8`s
+///
+/// The encoding of `T` must be equal to its binary representation, and size of `T` must be less
+/// or equal to [`MAX_PREALLOCATION`].
+fn read_vec_from_u8s<T, I>(input: &mut I, len: usize) -> Result<Vec<T>, Error>
+where
+	T: ToMutByteSlice + Default + Clone,
+	I: Input,
+{
+	let byte_len = len
+		.checked_mul(mem::size_of::<T>())
+		.ok_or("Item is too big and cannot be allocated")?;
 
-		// Here we pre-allocate only the maximum pre-allocation
-		let mut items: Vec<T> = vec![];
+	// Check if there is enough data in the input buffer.
+	if let Some(input_len) = input.remaining_len()? {
+		if input_len < byte_len {
+			return Err("Not enough data to decode vector".into());
+		}
+	}
 
-		let mut items_remains = items_len;
+	decode_vec_chunked(len, |decoded_vec, chunk_len| {
+		let decoded_vec_len = decoded_vec.len();
+		let decoded_vec_size = decoded_vec_len * mem::size_of::<T>();
+		unsafe {
+			decoded_vec.set_len(decoded_vec_len + chunk_len);
+		}
 
-		while items_remains > 0 {
-			let items_len_read = max_preallocated_items.min(items_remains);
+		let bytes_slice = decoded_vec.as_mut_byte_slice();
+		input.read(&mut bytes_slice[decoded_vec_size..])
+	})
+}
 
-			let items_len_filled = items.len();
-			let items_new_size = items_len_filled + items_len_read;
+fn decode_vec_from_items<T, I>(input: &mut I, len: usize) -> Result<Vec<T>, Error>
+where
+	T: Decode,
+	I: Input,
+{
+	// Check if there is enough data in the input buffer.
+	if let Some(input_len) = input.remaining_len()? {
+		if input_len < len {
+			return Err("Not enough data to decode vector".into());
+		}
+	}
```

Comment:
This isn't correct as deserializing `T` might take any number of bytes (including even zero bytes, e.g. `()`).

What we should do here is to have a `serialized_size_hint()` method (or, more specifically, probably an associated const so that it can be checked statically to fit within `MAX_PREALLOCATION`) or something like that on `T` which would return a value that could allow this check. (We already have `encoded_fixed_size` there, but that returns an *exact* number of bytes; it could be used here, but technically that's too strict and we can do better here by using the minimum.)

Reply:
Yeah, alternatively we can drop it. Although having it here can have one benefit - if we end up not having enough data then this will return an early error instead of wasting time trying to deserialize it. Nice to have, but not strictly necessary.

Reply:
To implement this, we would need to write quite a lot of code. For example for an `enum` we would need to know the variant that requires the least amount of bytes. However, it could then still fail at decoding because we try to decode always the enum variant that uses much more bytes etc.

Reply:
Hm, well, would it be *that* much code? I implemented this in my serialization crate and it's mostly fine; with `enum`s you essentially just autogenerate a `(min(variant1, variant2, ..), max(variant1, variant2, ..))` in your impl. Of course this is just an optimization (in some cases it would make incomplete deserializations fail early, and in some cases it would allow the compiler to remove per-element size checks), and as you've said it can still fail at decoding depending on what you're decoding.

Anyway, I'm fine with going with your suggestion to just delete the check.

Reply:
Dropped it: #615